Graph writing speeds

Graph writing appears to have slowed significantly on FreeBSD with the changes from #623 

With the current C++ siteupdate (5/21/2024), 8 threads, on FreeBSD, times for graph generation (computation and file writes):

Writing to the `/fast` SSD: ~86 sec
Writing to the `/home` magnetic drive: ~98 sec
Writing to `/ramdisk` RAM disk: ~96 sec

(only one run each for now)

But it doesn't seem to matter much where I write.  It also looks like the 8 threads are all pegging their cores during the graph generation process.

> FYI, I am planning to do an OS update for noreaster soon, at which time I'll also update all of the installed packages to their latest versions. I am just working to make sure I do so on a day when I can be at my office for several hours in case I break things to the point where I need physical access to the console.

Did this happen today?
Clang is generating pages of fatal errors due to not finding header files.
```sh
[yakra@noreaster ~/ytm/DataProcessing/siteupdate/cplusplus]$ gmake siteupdate
classes/GraphGeneration/HighwayGraphMT.d
In file included from classes/GraphGeneration/HighwayGraph.cpp:2:
In file included from classes/GraphGeneration/HighwayGraph.h:10:
In file included from /usr/include/c++/v1/list:201:
/usr/include/c++/v1/__algorithm/equal.h:20:10: fatal error: '__string/constexpr_c_functions.h' file not found
   20 | #include <__string/constexpr_c_functions.h>
      |          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
1 error generated.
```
```sh
[yakra@noreaster ~/ytm/DataProcessing/siteupdate/cplusplus]$ ls -l /usr/include/c++/v1/list
-r--r--r--  1 root  wheel  71793 May 29 15:17 /usr/include/c++/v1/list
```

--- a/SETUP.md
+++ b/SETUP.md
@@ -4,7 +4,7 @@ This document is intended to describe how one can set up an environment to run T
 
 Standard tools expected include `bash`, `bzip2`, a GNU version of `make` (often named `gmake`), `ssh`.
 
-There are two versions of the site update program: one written in Python and one in C++.  Both produce the same result.  The Python version requires a Python3 installation (as of this writing, Python 3.9.13).  Below, we will assume that Python can be launched with the command "python3".  The C++ version requires a C++ compiler (as of this writing, FreeBSD clang 13.0.0).
+There are two versions of the site update program: one written in Python and one in C++.  Both produce the same result.  The Python version requires a Python3 installation (as of this writing, Python 3.9.13).  Below, we will assume that Python can be launched with the command "python3".  The C++ version requires a C++ compiler (as of this writing, FreeBSD clang 13.0.0) and the [{fmt}](https://fmt.dev/) library (as of this writing, {fmt} 10.2.0).
 
 ### Cloning Needed Repositories
 

--- a/siteupdate/cplusplus/Makefile
+++ b/siteupdate/cplusplus/Makefile
@@ -4,7 +4,7 @@ ifeq ($(OS), Linux)
 endif
 
 CXX = clang++
-CXXFLAGS = -O3 -std=c++11 -Wno-comment -Wno-dangling-else -Wno-logical-op-parentheses
+CXXFLAGS = -O3 -std=c++11 -isystem /usr/local/include -Wno-comment -Wno-dangling-else -Wno-logical-op-parentheses
 
 STObjects = \
   classes/GraphGeneration/HighwayGraphST.o \

--- a/siteupdate/cplusplus/classes/Args/Args.cpp
+++ b/siteupdate/cplusplus/classes/Args/Args.cpp
@@ -6,7 +6,7 @@
 /* e */ bool Args::errorcheck = 0;
 /* k */ bool Args::skipgraphs = 0;
 /* v */ bool Args::mtvertices = 0;
-/* C */ bool Args::mtcsvfiles = 0;
+/* C */ bool Args::stcsvfiles = 0;
 /* E */ bool Args::edgecounts = 0;
 /* b */ bool Args::bitsetlogs = 0;
 /* w */ std::string Args::datapath = "../../HighwayData";
@@ -34,7 +34,7 @@ bool Args::init(int argc, char *argv[])
 	{	     if ARG(0, "-e", "--errorcheck")		 errorcheck = 1;
 		else if ARG(0, "-k", "--skipgraphs")		 skipgraphs = 1;
 		else if ARG(0, "-v", "--mt-vertices")		 mtvertices = 1;
-		else if ARG(0, "-C", "--mt-csvs")		 mtcsvfiles = 1;
+		else if ARG(0, "-C", "--st-csvs")		 stcsvfiles = 1;
 		else if ARG(0, "-E", "--edge-counts")		 edgecounts = 1;
 		else if ARG(0, "-b", "--bitset-logs")		 bitsetlogs = 1;
 		else if ARG(0, "-h", "--help")			{show_help(); return 1;}
@@ -127,7 +127,7 @@ void Args::show_help()
 	std::cout  <<  "		        Number of digits (1-9) after decimal point in\n";
 	std::cout  <<  "		        timestamp readouts\n";
 	std::cout  <<  "  -v, --mt-vertices     Multi-threaded vertex construction\n";
-	std::cout  <<  "  -C, --mt-csvs         Multi-threaded stats csv files\n";
+	std::cout  <<  "  -C, --st-csvs         Single-threaded stats csv files\n";
 	std::cout  <<  "  -E, --edge-counts     Report the quantity of each format graph edge\n";
 	std::cout  <<  "  -b, --bitset-logs     Write TMBitset RAM use logs for region & system\n";
 	std::cout  <<  "		        vertices & edges\n";

--- a/siteupdate/cplusplus/classes/Args/Args.h
+++ b/siteupdate/cplusplus/classes/Args/Args.h
@@ -18,7 +18,7 @@ class Args
 	/* e */ static bool errorcheck;
 	/* T */ static int timeprecision;
 	/* v */ static bool mtvertices;
-	/* C */ static bool mtcsvfiles;
+	/* C */ static bool stcsvfiles;
 	/* E */ static bool edgecounts;
 	/* b */ static bool bitsetlogs;
 	/* L */ static int colocationlimit;

--- a/siteupdate/cplusplus/classes/GraphGeneration/HGEdge.cpp
+++ b/siteupdate/cplusplus/classes/GraphGeneration/HGEdge.cpp
@@ -1,3 +1,4 @@
+#define FMT_HEADER_ONLY
 #include "HGEdge.h"
 #include "HGVertex.h"
 #include "../Args/Args.h"
@@ -6,6 +7,7 @@
 #include "../Route/Route.h"
 #include "../Waypoint/Waypoint.h"
 #include "../../templates/contains.cpp"
+#include <fmt/format.h>
 
 HGVertex* HGEdge::v_array;
 
@@ -107,21 +109,25 @@ void HGEdge::detach()
 }
 
 // write line to tmg collapsed edge file
-void HGEdge::collapsed_tmg_line(std::ofstream& file, unsigned int threadnum, std::vector<HighwaySystem*> *systems)
+void HGEdge::collapsed_tmg_line(std::ofstream& file, char* fstr, unsigned int threadnum, std::vector<HighwaySystem*> *systems)
 {	file << vertex1->c_vertex_num[threadnum] << ' ' << vertex2->c_vertex_num[threadnum] << ' ';
 	segment->write_label(file, systems);
 	for (HGVertex *intermediate : intermediate_points)
-		file << ' ' << intermediate->lat << ' ' << intermediate->lng;
+	{	*fmt::format_to(fstr, " {:.15} {:.15}", intermediate->lat, intermediate->lng) = 0;
+		file << fstr;
+	}
 	file << '\n';
 }
 
 // write line to tmg traveled edge file
-void HGEdge::traveled_tmg_line(std::ofstream& file, unsigned int threadnum, std::vector<HighwaySystem*> *systems, bool trav, char* code)
+void HGEdge::traveled_tmg_line(std::ofstream& file, char* fstr, unsigned int threadnum, std::vector<HighwaySystem*> *systems, bool trav, char* code)
 {	file << vertex1->t_vertex_num[threadnum] << ' ' << vertex2->t_vertex_num[threadnum] << ' ';
 	segment->write_label(file, systems);
 	file << ' ' << (trav ? segment->clinchedby_code(code, threadnum) : "0");
 	for (HGVertex *intermediate : intermediate_points)
-		file << ' ' << intermediate->lat << ' ' << intermediate->lng;
+	{	*fmt::format_to(fstr, " {:.15} {:.15}", intermediate->lat, intermediate->lng) = 0;
+		file << fstr;
+	}
 	file << '\n';
 }
 
@@ -131,7 +137,7 @@ std::string HGEdge::debug_tmg_line(std::vector<HighwaySystem*> *systems, unsigne
 			 + std::to_string(vertex2->c_vertex_num[threadnum]) + " [" + *vertex2->unique_name + "] " + label(systems);
 	char fstr[58];
 	for (HGVertex *intermediate : intermediate_points)
-	{	sprintf(fstr, "] %.15g %.15g", intermediate->lat, intermediate->lng);
+	{	*fmt::format_to(fstr, "] {:.15} {:.15}", intermediate->lat, intermediate->lng) = 0;
 		line += " [" + *intermediate->unique_name + fstr;
 	}
 	return line;
@@ -157,7 +163,7 @@ std::string HGEdge::intermediate_point_string()
 	std::string line = "";
 	char fstr[56];
 	for (HGVertex *i : intermediate_points)
-	{	sprintf(fstr, "%.15g %.15g", i->lat, i->lng);
+	{	*fmt::format_to(fstr, "{:.15} {:.15}", i->lat, i->lng) = 0;
 		line += " [" + *i->unique_name + "] " + fstr;
 	}
 	return line;

--- a/siteupdate/cplusplus/classes/GraphGeneration/HGEdge.h
+++ b/siteupdate/cplusplus/classes/GraphGeneration/HGEdge.h
@@ -31,8 +31,8 @@ class HGEdge
 	HGEdge(HGVertex *, unsigned char, HGEdge*, HGEdge*);
 
 	void detach();
-	void collapsed_tmg_line(std::ofstream&, unsigned int, std::vector<HighwaySystem*>*);
-	void traveled_tmg_line (std::ofstream&, unsigned int, std::vector<HighwaySystem*>*, bool, char*);
+	void collapsed_tmg_line(std::ofstream&, char*, unsigned int, std::vector<HighwaySystem*>*);
+	void traveled_tmg_line (std::ofstream&, char*, unsigned int, std::vector<HighwaySystem*>*, bool, char*);
 	std::string debug_tmg_line(std::vector<HighwaySystem*> *, unsigned int);
 	std::string str();
 	std::string intermediate_point_string();

--- a/siteupdate/cplusplus/classes/GraphGeneration/HighwayGraph.cpp
+++ b/siteupdate/cplusplus/classes/GraphGeneration/HighwayGraph.cpp
@@ -1,3 +1,4 @@
+#define FMT_HEADER_ONLY
 #include "HighwayGraph.h"
 #include "GraphListEntry.h"
 #include "HGEdge.h"
@@ -13,6 +14,7 @@
 #include "../Waypoint/Waypoint.h"
 #include "../WaypointQuadtree/WaypointQuadtree.h"
 #include "../../templates/contains.cpp"
+#include <fmt/format.h>
 #include <fstream>
 #include <thread>
 
@@ -345,9 +347,9 @@ void HighwayGraph::bitsetlogs(HGVertex* hp_end)
 //     for intermediate "shaping points" along the edge, ordered from endpoint 1 to endpoint 2.
 //
 void HighwayGraph::write_master_graphs_tmg()
-{	std::ofstream simplefile(Args::graphfilepath + "/tm-master-simple.tmg");   simplefile.precision(15);
-	std::ofstream collapfile(Args::graphfilepath + "/tm-master.tmg");	   collapfile.precision(15);
-	std::ofstream travelfile(Args::graphfilepath + "/tm-master-traveled.tmg"); travelfile.precision(15);
+{	std::ofstream simplefile(Args::graphfilepath + "/tm-master-simple.tmg");
+	std::ofstream collapfile(Args::graphfilepath + "/tm-master.tmg");
+	std::ofstream travelfile(Args::graphfilepath + "/tm-master-traveled.tmg");
 	simplefile << "TMG 1.0 simple\n";
 	collapfile << "TMG 1.0 collapsed\n";
 	travelfile << "TMG 2.0 traveled\n";
@@ -359,14 +361,13 @@ void HighwayGraph::write_master_graphs_tmg()
 	unsigned int sv = 0;
 	unsigned int cv = 0;
 	unsigned int tv = 0;
+	char fstr[57];
 	for (HGVertex& v : vertices)
-	{	auto write_vertex = [&](std::ofstream& tmg)
-		{	tmg << *(v.unique_name) << ' ' << v.lat << ' ' << v.lng << '\n';
-		};
+	{	*fmt::format_to(fstr, " {:.15} {:.15}", v.lat, v.lng) = 0;
 		switch (v.visibility) // fall-thru is a Good Thing!
-		{ case 2:  write_vertex(collapfile); v.c_vertex_num[0] = cv++;
-		  case 1:  write_vertex(travelfile); v.t_vertex_num[0] = tv++;
-		  default: write_vertex(simplefile); v.s_vertex_num[0] = sv++;
+		{ case 2:  collapfile << *(v.unique_name) << fstr << '\n'; v.c_vertex_num[0] = cv++;
+		  case 1:  travelfile << *(v.unique_name) << fstr << '\n'; v.t_vertex_num[0] = tv++;
+		  default: simplefile << *(v.unique_name) << fstr << '\n'; v.s_vertex_num[0] = sv++;
 		}
 	}
 
@@ -380,10 +381,10 @@ void HighwayGraph::write_master_graphs_tmg()
 	//TODO: multiple functions performing the same instructions for multiple files?
 	for (HGEdge *e = edges.begin(), *end = edges.end(); e != end; ++e)
 	{ if (e->format & HGEdge::collapsed)
-		e->collapsed_tmg_line(collapfile, 0, 0);
+		e->collapsed_tmg_line(collapfile, fstr, 0, 0);
 	  if (e->format & HGEdge::traveled)
 	  {	for (char*n=cbycode; n<cbycode+nibbles; ++n) *n = '0';
-		e->traveled_tmg_line(travelfile, 0, 0, TravelerList::allusers.size, cbycode);
+		e->traveled_tmg_line(travelfile, fstr, 0, 0, TravelerList::allusers.size, cbycode);
 	  }
 	  if (e->format & HGEdge::simple)
 	  {	simplefile << e->vertex1->s_vertex_num[0] << ' '
@@ -418,9 +419,9 @@ void HighwayGraph::write_subgraphs_tmg
 {	unsigned int cv_count = 0, sv_count = 0, tv_count = 0;
 	unsigned int ce_count = 0, se_count = 0, te_count = 0;
 	GraphListEntry* g = GraphListEntry::entries.data()+graphnum;
-	std::ofstream simplefile(Args::graphfilepath+'/'+g -> filename()); simplefile.precision(15);
-	std::ofstream collapfile(Args::graphfilepath+'/'+g[1].filename()); collapfile.precision(15);
-	std::ofstream travelfile(Args::graphfilepath+'/'+g[2].filename()); travelfile.precision(15);
+	std::ofstream simplefile(Args::graphfilepath+'/'+g -> filename());
+	std::ofstream collapfile(Args::graphfilepath+'/'+g[1].filename());
+	std::ofstream travelfile(Args::graphfilepath+'/'+g[2].filename());
 	TMBitset<HGVertex*, uint64_t> mv; // vertices matching all criteria
 	TMBitset<HGEdge*,   uint64_t> me; //    edges matching all criteria
 	std::vector<TravelerList*> traveler_lists;
@@ -455,14 +456,13 @@ void HighwayGraph::write_subgraphs_tmg
 	unsigned int sv = 0;
 	unsigned int cv = 0;
 	unsigned int tv = 0;
+	char fstr[57];
 	for (HGVertex *v : mv)
-	{	auto write_vertex = [&](std::ofstream& tmg)
-		{	tmg << *(v->unique_name) << ' ' << v->lat << ' ' << v->lng << '\n';
-		};
+	{	*fmt::format_to(fstr, " {:.15} {:.15}", v->lat, v->lng) = 0;
 		switch(v->visibility) // fall-thru is a Good Thing!
-		{ case 2:  write_vertex(collapfile); v->c_vertex_num[threadnum] = cv++;
-		  case 1:  write_vertex(travelfile); v->t_vertex_num[threadnum] = tv++;
-		  default: write_vertex(simplefile); v->s_vertex_num[threadnum] = sv++;
+		{ case 2:  collapfile << *(v->unique_name) << fstr << '\n'; v->c_vertex_num[threadnum] = cv++;
+		  case 1:  travelfile << *(v->unique_name) << fstr << '\n'; v->t_vertex_num[threadnum] = tv++;
+		  default: simplefile << *(v->unique_name) << fstr << '\n'; v->s_vertex_num[threadnum] = sv++;
 		}
 	}
 
@@ -481,10 +481,10 @@ void HighwayGraph::write_subgraphs_tmg
 		simplefile << '\n';
 	  }
 	  if (e->format & HGEdge::collapsed)
-		e->collapsed_tmg_line(collapfile, threadnum, g->systems);
+		e->collapsed_tmg_line(collapfile, fstr, threadnum, g->systems);
 	  if (e->format & HGEdge::traveled)
 	  {	for (char*n=cbycode; n<cbycode+nibbles; ++n) *n = '0';
-		e->traveled_tmg_line (travelfile, threadnum, g->systems, travnum, cbycode);
+		e->traveled_tmg_line (travelfile, fstr, threadnum, g->systems, travnum, cbycode);
 	  }
 	}
 	delete[] cbycode;

--- a/siteupdate/cplusplus/classes/HighwaySystem/HighwaySystem.cpp
+++ b/siteupdate/cplusplus/classes/HighwaySystem/HighwaySystem.cpp
@@ -1,3 +1,4 @@
+#define FMT_HEADER_ONLY
 #include "HighwaySystem.h"
 #include "../Args/Args.h"
 #include "../ConnectedRoute/ConnectedRoute.h"
@@ -11,6 +12,7 @@
 #include "../TravelerList/TravelerList.h"
 #include "../Waypoint/Waypoint.h"
 #include "../../functions/tmstring.h"
+#include <fmt/format.h>
 #include <fstream>
 
 TMArray<HighwaySystem> HighwaySystem::syslist;
@@ -219,24 +221,25 @@ void HighwaySystem::stats_csv()
 	sysfile << '\n';
 	for (TravelerList& t : TravelerList::allusers)
 	{	// only include entries for travelers who have any mileage in system
-		auto it = t.system_region_mileages.find(this);
-		if (it != t.system_region_mileages.end())
-		{	sprintf(fstr, ",%.2f", t.system_miles(this));
+		auto srm_it = t.system_region_mileages.find(this);
+		if (srm_it != t.system_region_mileages.end())
+		{	*fmt::format_to(fstr, ",{:.2f}", t.system_miles(this)) = 0;
 			sysfile << t.traveler_name << fstr;
 			for (Region *region : regions)
-			  try {	sprintf(fstr, ",%.2f", it->second.at(region));
-				sysfile << fstr;
-			      }
-			  catch (const std::out_of_range& oor)
-			      {	sysfile << ",0";
-			      }
+			{	auto trm_it = srm_it->second.find(region);
+				if (trm_it != srm_it->second.end())
+				{	*fmt::format_to(fstr, ",{:.2f}", trm_it->second) = 0;
+					sysfile << fstr;
+				}
+				else	sysfile << ",0";
+			}
 			sysfile << '\n';
 		}
 	}
-	sprintf(fstr, "TOTAL,%.2f", total_mileage());
+	*fmt::format_to(fstr, "TOTAL,{:.2f}", total_mileage()) = 0;
 	sysfile << fstr;
 	for (Region *region : regions)
-	{	sprintf(fstr, ",%.2f", mileage_by_region.at(region));
+	{	*fmt::format_to(fstr, ",{:.2f}", mileage_by_region.at(region)) = 0;
 		sysfile << fstr;
 	}
 	sysfile << '\n';

--- a/siteupdate/cplusplus/classes/TravelerList/userlog.cpp
+++ b/siteupdate/cplusplus/classes/TravelerList/userlog.cpp
@@ -1,3 +1,4 @@
+#define FMT_HEADER_ONLY
 #include "TravelerList.h"
 #include "../Args/Args.h"
 #include "../ConnectedRoute/ConnectedRoute.h"
@@ -5,26 +6,16 @@
 #include "../Region/Region.h"
 #include "../Route/Route.h"
 #include "../../functions/tmstring.h"
+#include <fmt/format.h>
 #include <fstream>
 
 void TravelerList::userlog(const double total_active_only_miles, const double total_active_preview_miles)
-{	std::cout << "." << std::flush;
+{	char fstr[112];
+	std::cout << "." << std::flush;
 	std::ofstream log(Args::logfilepath+"/users/"+traveler_name+".log", std::ios::app);
-	log.setf(std::ios::fixed);
-	log.precision(2);
-	auto log_clinched_mi = [&](double clinched, double total)->std::ofstream&
-	{	log << clinched << " of " << total << " mi ";
-		if (total)
-			log << "(" << 100*clinched/total << "%)";
-		else	log << "-.--%";
-		return	log;
-	};
-
 	log << "Clinched Highway Statistics\n";
-	log << "Overall in active systems: ";
-	log_clinched_mi(active_only_miles(), total_active_only_miles) << '\n';
-	log << "Overall in active+preview systems: ";
-	log_clinched_mi(active_preview_miles(), total_active_preview_miles) << '\n';
+	log << "Overall in active systems: " << format_clinched_mi(fstr, active_only_miles(), total_active_only_miles) << '\n';
+	log << "Overall in active+preview systems: " << format_clinched_mi(fstr, active_preview_miles(), total_active_preview_miles) << '\n';
 
 	log << "Overall by region: (each line reports active only then active+preview)\n";
 	std::list<Region*> travregions;
@@ -35,9 +26,8 @@ void TravelerList::userlog(const double total_active_only_miles, const double to
 	{	double t_active_miles = 0;
 		if (active_only_mileage_by_region.count(region))
 			t_active_miles = active_only_mileage_by_region.at(region);
-		log << region->code << ": ";
-		log_clinched_mi(t_active_miles, region->active_only_mileage) << ", ";
-		log_clinched_mi(active_preview_mileage_by_region.at(region), region->active_preview_mileage) << '\n';
+		log << region->code << ": " << format_clinched_mi(fstr, t_active_miles, region->active_only_mileage) << ", "
+		    << format_clinched_mi(fstr, active_preview_mileage_by_region.at(region), region->active_preview_mileage) << '\n';
 	}
 	unsigned int active_systems_traveled = 0;
 	unsigned int active_systems_clinched = 0;
@@ -58,9 +48,8 @@ void TravelerList::userlog(const double total_active_only_miles, const double to
 			// the DB, but add to logs only if it's been traveled at
 			// all and it covers multiple regions
 			auto& sysmbr = h->mileage_by_region;
-			log.precision(2);
-			log << "System " << h->systemname << " (" << h->level_name() << ") overall: ";
-			log_clinched_mi(t_system_overall, h->total_mileage()) << '\n';
+			log << "System " << h->systemname << " (" << h->level_name() << ") overall: "
+			    << format_clinched_mi(fstr, t_system_overall, h->total_mileage()) << '\n';
 			if (sysmbr.size() > 1)
 			{	log << "System " << h->systemname << " by region:\n";
 				std::list<Region*> sysregions;
@@ -72,8 +61,7 @@ void TravelerList::userlog(const double total_active_only_miles, const double to
 					auto it = system_region_mileages.at(h).find(region);
 					if (it != system_region_mileages.at(h).end())
 						system_region_mileage = it->second;
-					log << "  " << region->code << ": ";
-					log_clinched_mi(system_region_mileage, sysmbr.at(region)) << '\n';
+					log << "  " << region->code << ": " << format_clinched_mi(fstr, system_region_mileage, sysmbr.at(region)) << '\n';
 				}
 			}
 
@@ -99,14 +87,12 @@ void TravelerList::userlog(const double total_active_only_miles, const double to
 				{	num_con_rtes_traveled += 1;
 					num_con_rtes_clinched += (con_clinched_miles == cr.mileage);
 					ccr_values.emplace_back(&cr, con_clinched_miles);
-					log << cr.readable_name() << ": ";
-					log_clinched_mi(con_clinched_miles, cr.mileage) << '\n';
+					log << cr.readable_name() << ": " << format_clinched_mi(fstr, con_clinched_miles, cr.mileage) << '\n';
 					if (roots.size() == 1)
 						log << " (" << roots[0]->readable_name() << " only)\n";
 					else {	for (auto& rm : chop_mi)
-						{   log << "  " << rm.first->readable_name() << ": ";
-						    log_clinched_mi(rm.second,rm.first->mileage) << "\n";
-						}
+						    log << "  " << rm.first->readable_name() << ": "
+							<< format_clinched_mi(fstr, rm.second,rm.first->mileage) << "\n";
 						log << '\n';
 					     }
 				}
@@ -115,26 +101,23 @@ void TravelerList::userlog(const double total_active_only_miles, const double to
 			  if (h->active())
 				active_systems_clinched++;
 			  else	preview_systems_clinched++;
-			log.precision(1);
-			log << "System " << h->systemname
-			    << " connected routes traveled: " << num_con_rtes_traveled << " of " << h->con_routes.size
-			    << " (" << 100*(double)num_con_rtes_traveled/h->con_routes.size
-					  << "%), clinched: " << num_con_rtes_clinched << " of " << h->con_routes.size
-			    << " (" << 100*(double)num_con_rtes_clinched/h->con_routes.size << "%).\n";
+			*fmt::format_to(fstr, " connected routes traveled: {} of {} ({:.1f}%), clinched: {} of {} ({:.1f}%).",
+				num_con_rtes_traveled, (int)h->con_routes.size, 100*(double)num_con_rtes_traveled/h->con_routes.size,
+				num_con_rtes_clinched, (int)h->con_routes.size, 100*(double)num_con_rtes_clinched/h->con_routes.size) = 0;
+			log << "System " << h->systemname << fstr << '\n';
 		}
 	  }
-	log.precision(1); // for the few users with no traveled mileage
 
 	// grand summary, active only
-	log << "\nTraveled " << active_systems_traveled << " of " << HighwaySystem::num_active
-		      << " (" << 100*(double)active_systems_traveled/HighwaySystem::num_active
-	  << "%), Clinched " << active_systems_clinched << " of " << HighwaySystem::num_active
-		      << " (" << 100*(double)active_systems_clinched/HighwaySystem::num_active << "%) active systems\n";
+	*fmt::format_to(fstr,"\nTraveled {} of {} ({:.1f}%), Clinched {} of {} ({:.1f}%) active systems",
+		active_systems_traveled, HighwaySystem::num_active, 100*(double)active_systems_traveled/HighwaySystem::num_active,
+		active_systems_clinched, HighwaySystem::num_active, 100*(double)active_systems_clinched/HighwaySystem::num_active) = 0;
+	log << fstr << '\n';
 	// grand summary, active+preview
-	log << "Traveled " << preview_systems_traveled << " of " << HighwaySystem::num_preview
-		    << " (" << 100*(double)preview_systems_traveled/HighwaySystem::num_preview
-	<< "%), Clinched " << preview_systems_clinched << " of " << HighwaySystem::num_preview
-		    << " (" << 100*(double)preview_systems_clinched/HighwaySystem::num_preview << "%) preview systems\n";
+	*fmt::format_to(fstr,"Traveled {} of {} ({:.1f}%), Clinched {} of {} ({:.1f}%) preview systems",
+		preview_systems_traveled, HighwaySystem::num_preview, 100*(double)preview_systems_traveled/HighwaySystem::num_preview,
+		preview_systems_clinched, HighwaySystem::num_preview, 100*(double)preview_systems_clinched/HighwaySystem::num_preview) = 0;
+	log << fstr << '\n';
 
 	// updated routes, sorted by date
 	log << "\nMost recent updates for listed routes:\n";

--- a/siteupdate/cplusplus/functions/allbyregionactiveonly.cpp
+++ b/siteupdate/cplusplus/functions/allbyregionactiveonly.cpp
@@ -1,7 +1,9 @@
+#define FMT_HEADER_ONLY
 #include "../classes/Args/Args.h"
 #include "../classes/Region/Region.h"
 #include "../classes/TravelerList/TravelerList.h"
 #include "../threads/threads.h"
+#include <fmt/format.h>
 #include <fstream>
 
 void allbyregionactiveonly(std::mutex* mtx, double total_mi)
@@ -19,21 +21,22 @@ void allbyregionactiveonly(std::mutex* mtx, double total_mi)
 	{	double t_total_mi = 0;
 		for (std::pair<Region* const, double>& rm : t.active_only_mileage_by_region)
 			t_total_mi += rm.second;
-		sprintf(fstr, "%.2f", t_total_mi);
+		*fmt::format_to(fstr, "{:.2f}", t_total_mi) = 0;
 		allfile << t.traveler_name << ',' << fstr;
 		for (Region *region : regions)
-		  try {	sprintf(fstr, "%.2f", t.active_only_mileage_by_region.at(region));
-			allfile << ',' << fstr;
-		      }
-		  catch (const std::out_of_range& oor)
-		      {	allfile << ",0";
-		      }
+		{	auto it = t.active_only_mileage_by_region.find(region);
+			if (it != t.active_only_mileage_by_region.end())
+			{	*fmt::format_to(fstr, "{:.2f}", it->second) = 0;
+				allfile << ',' << fstr;
+			}
+		  	else	allfile << ",0";
+		}
 		allfile << '\n';
 	}
-	sprintf(fstr, "TOTAL,%.2f", total_mi);
+	*fmt::format_to(fstr, "TOTAL,{:.2f}", total_mi) = 0;
 	allfile << fstr;
 	for (Region *region : regions)
-	{	sprintf(fstr, ",%.2f", region->active_only_mileage);
+	{	*fmt::format_to(fstr, ",{:.2f}", region->active_only_mileage) = 0;
 		allfile << fstr;
 	}
 	allfile << '\n';

--- a/siteupdate/cplusplus/functions/allbyregionactivepreview.cpp
+++ b/siteupdate/cplusplus/functions/allbyregionactivepreview.cpp
@@ -1,7 +1,9 @@
+#define FMT_HEADER_ONLY
 #include "../classes/Args/Args.h"
 #include "../classes/Region/Region.h"
 #include "../classes/TravelerList/TravelerList.h"
 #include "../threads/threads.h"
+#include <fmt/format.h>
 #include <fstream>
 
 void allbyregionactivepreview(std::mutex* mtx, double total_mi)
@@ -19,21 +21,22 @@ void allbyregionactivepreview(std::mutex* mtx, double total_mi)
 	{	double t_total_mi = 0;
 		for (std::pair<Region* const, double>& rm : t.active_preview_mileage_by_region)
 			t_total_mi += rm.second;
-		sprintf(fstr, "%.2f", t_total_mi);
+		*fmt::format_to(fstr, "{:.2f}", t_total_mi) = 0;
 		allfile << t.traveler_name << ',' << fstr;
 		for (Region *region : regions)
-		  try {	sprintf(fstr, "%.2f", t.active_preview_mileage_by_region.at(region));
-			allfile << ',' << fstr;
-		      }
-		  catch (const std::out_of_range& oor)
-		      {	allfile << ",0";
-		      }
+		{	auto it = t.active_preview_mileage_by_region.find(region);
+			if (it != t.active_preview_mileage_by_region.end())
+			{	*fmt::format_to(fstr, "{:.2f}", it->second) = 0;
+				allfile << ',' << fstr;
+			}
+			else	allfile << ",0";
+		}
 		allfile << '\n';
 	}
-	sprintf(fstr, "TOTAL,%.2f", total_mi);
+	*fmt::format_to(fstr, "TOTAL,{:.2f}", total_mi) = 0;
 	allfile << fstr;
 	for (Region *region : regions)
-	{	sprintf(fstr, ",%.2f", region->active_preview_mileage);
+	{	*fmt::format_to(fstr, ",{:.2f}", region->active_preview_mileage) = 0;
 		allfile << fstr;
 	}
 	allfile << '\n';

--- a/siteupdate/cplusplus/functions/rdstats.cpp
+++ b/siteupdate/cplusplus/functions/rdstats.cpp
@@ -1,8 +1,10 @@
+#define FMT_HEADER_ONLY
 #include "../classes/Args/Args.h"
 #include "../classes/ConnectedRoute/ConnectedRoute.h"
 #include "../classes/HighwaySystem/HighwaySystem.h"
 #include "../classes/Region/Region.h"
 #include "../classes/Route/Route.h"
+#include <fmt/format.h>
 #include <fstream>
 #include <list>
 
@@ -18,24 +20,24 @@ void rdstats(double& active_only_miles, double& active_preview_miles, time_t* ti
 		active_preview_miles += r.active_preview_mileage;
 		overall_miles += r.overall_mileage;
 	}
-	sprintf(fstr, "Active routes (active): %.2f mi\n", active_only_miles);
+	*fmt::format_to(fstr, "Active routes (active): {:.2f} mi\n", active_only_miles) = 0;
 	rdstatsfile << fstr;
-	sprintf(fstr, "Clinchable routes (active, preview): %.2f mi\n", active_preview_miles);
+	*fmt::format_to(fstr, "Clinchable routes (active, preview): {:.2f} mi\n", active_preview_miles) = 0;
 	rdstatsfile << fstr;
-	sprintf(fstr, "All routes (active, preview, devel): %.2f mi\n", overall_miles);
+	*fmt::format_to(fstr, "All routes (active, preview, devel): {:.2f} mi\n", overall_miles) = 0;
 	rdstatsfile << fstr;
 	rdstatsfile << "Breakdown by region:\n";
 	// a nice enhancement later here might break down by continent, then country,
 	// then region
 	for (Region& region : Region::allregions)
 	  if (region.overall_mileage)
-	  {	sprintf(fstr, ": %.2f (active), %.2f (active, preview) %.2f (active, preview, devel)\n",
-			region.active_only_mileage, region.active_preview_mileage, region.overall_mileage);
+	  {	*fmt::format_to(fstr, ": {:.2f} (active), {:.2f} (active, preview) {:.2f} (active, preview, devel)\n",
+				region.active_only_mileage, region.active_preview_mileage, region.overall_mileage) = 0;
 		rdstatsfile << region.code << fstr;
 	  }
 
 	for (HighwaySystem& h : HighwaySystem::syslist)
-	{	sprintf(fstr, ") total: %.2f mi\n", h.total_mileage());
+	{	*fmt::format_to(fstr, ") total: {:.2f} mi\n", h.total_mileage()) = 0;
 		rdstatsfile << "System " << h.systemname << " (" << h.level_name() << fstr;
 		if (h.mileage_by_region.size() > 1)
 		{	rdstatsfile << "System " << h.systemname << " by region:\n";
@@ -44,7 +46,7 @@ void rdstats(double& active_only_miles, double& active_preview_miles, time_t* ti
 				regions_in_system.push_back(rm.first);
 			regions_in_system.sort();
 			for (Region *r : regions_in_system)
-			{	sprintf(fstr, ": %.2f mi\n", h.mileage_by_region.at(r));
+			{	*fmt::format_to(fstr, ": {:.2f} mi\n", h.mileage_by_region.at(r)) = 0;
 				rdstatsfile << r->code << fstr;
 			}
 		}
@@ -52,11 +54,11 @@ void rdstats(double& active_only_miles, double& active_preview_miles, time_t* ti
 		for (ConnectedRoute& cr : h.con_routes)
 		{	std::string to_write = "";
 			for (Route *r : cr.roots)
-			{	sprintf(fstr, ": %.2f mi\n", r->mileage);
+			{	*fmt::format_to(fstr, ": {:.2f} mi\n", r->mileage) = 0;
 				to_write += "  " + r->readable_name() + fstr;
 				cr.mileage += r->mileage;
 			}
-			sprintf(fstr, ": %.2f mi", cr.mileage);
+			*fmt::format_to(fstr, ": {:.2f} mi", cr.mileage) = 0;
 			rdstatsfile << cr.readable_name() << fstr;
 			if (cr.roots.size() == 1)
 				rdstatsfile << " (" << cr.roots[0]->readable_name() << " only)\n";

--- a/siteupdate/cplusplus/functions/sql_file.cpp
+++ b/siteupdate/cplusplus/functions/sql_file.cpp
@@ -1,3 +1,4 @@
+#define FMT_HEADER_ONLY
 #include "sql_file.h"
 #include "tmstring.h"
 #include "../classes/Args/Args.h"
@@ -12,6 +13,7 @@
 #include "../classes/Route/Route.h"
 #include "../classes/TravelerList/TravelerList.h"
 #include "../classes/Waypoint/Waypoint.h"
+#include <fmt/format.h>
 #include <fstream>
 
 void sqlfile1
@@ -19,10 +21,10 @@ void sqlfile1
 	std::list<std::string*> *updates,
 	std::list<std::string*> *systemupdates,
 	std::mutex* term_mtx
-    ){	// Once all data is read in and processed, create a .sql file that will
+    ){	char fstr[65];
+	// Once all data is read in and processed, create a .sql file that will
 	// create all of the DB tables to be used by other parts of the project
 	std::ofstream sqlfile(Args::databasename+".sql");
-	sqlfile.precision(17);
 	// Note: removed "USE" line, DB name must be specified on the mysql command line
 
 	// we have to drop tables in the right order to avoid foreign key errors
@@ -138,8 +140,9 @@ void sqlfile1
 	  for (Route& r : h.routes)
 	  {	if (!first) sqlfile << ',';
 		first = 0;
+		*fmt::format_to(fstr, "{}", r.mileage) = 0;
 		sqlfile << "('" << r.system->systemname << "','" << r.region->code << "','" << r.route << "','" << r.banner << "','" << r.abbrev
-			<< "','" << double_quotes(r.city) << "','" << r.root << "','" << r.mileage << "','" << r.rootOrder << "','" << csvOrder << "')\n";
+			<< "','" << double_quotes(r.city) << "','" << r.root << "','" << fstr << "','" << r.rootOrder << "','" << csvOrder << "')\n";
 		csvOrder += 1;
 	  }
 	sqlfile << ";\n";
@@ -161,8 +164,9 @@ void sqlfile1
 	  for (ConnectedRoute& cr : h.con_routes)
 	  {	if (!first) sqlfile << ',';
 		first = 0;
-		sqlfile << "('" << cr.system->systemname << "','" << cr.route << "','" << cr.banner << "','" << double_quotes(cr.groupname) << "','"
-			<< (cr.roots.size() ? cr.roots[0]->root.data() : "ERROR_NO_ROOTS") << "','" << cr.mileage << "','" << csvOrder << "')\n";
+		*fmt::format_to(fstr, "','{}'", cr.mileage) = 0;
+		sqlfile << "('" << cr.system->systemname << "','" << cr.route << "','" << cr.banner << "','" << double_quotes(cr.groupname)
+			<< "','" << (cr.roots.size() ? cr.roots[0]->root.data() : "ERROR_NO_ROOTS") << fstr << ",'" << csvOrder << "')\n";
 		csvOrder += 1;
 	  }
 	sqlfile << ";\n";
@@ -190,7 +194,6 @@ void sqlfile1
       #ifndef threading_enabled
 	std::cout << et->et() << "...waypoints" << std::endl;
       #endif
-	sqlfile.precision(15);
 	sqlfile << "CREATE TABLE waypoints (pointId INTEGER, pointName VARCHAR(" << DBFieldLength::label
 		<< "), latitude DOUBLE, longitude DOUBLE, root VARCHAR(" << DBFieldLength::root
 		<< "), PRIMARY KEY(pointId), FOREIGN KEY (root) REFERENCES routes(root));\n";
@@ -203,7 +206,8 @@ void sqlfile1
 		{	if (!first) sqlfile << ',';
 			first = 0;
 			sqlfile << "('" << point_num << "','" << w.label;
-			sqlfile << "','" << w.lat << "','" << w.lng;
+			*fmt::format_to(fstr,"','{:.15}",w.lat) = 0; sqlfile << fstr; if (w.lat==int(w.lat)) sqlfile << ".0";
+			*fmt::format_to(fstr,"','{:.15}",w.lng) = 0; sqlfile << fstr; if (w.lng==int(w.lng)) sqlfile << ".0";
 			sqlfile << "','" << r.root + "')\n";
 			point_num+=1;
 		}
@@ -218,7 +222,6 @@ void sqlfile1
       #ifndef threading_enabled
 	std::cout << et->et() << "...segments" << std::endl;
       #endif
-	sqlfile.precision(17);
 	sqlfile << "CREATE TABLE segments (segmentId INTEGER, waypoint1 INTEGER, waypoint2 INTEGER, root VARCHAR(" << DBFieldLength::root
 		<< "), PRIMARY KEY (segmentId), FOREIGN KEY (waypoint1) REFERENCES waypoints(pointId), "
 		<< "FOREIGN KEY (waypoint2) REFERENCES waypoints(pointId), FOREIGN KEY (root) REFERENCES routes(root));\n";
@@ -274,7 +277,8 @@ void sqlfile1
 	{	if (region.active_only_mileage+region.active_preview_mileage == 0) continue;
 		if (!first) sqlfile << ',';
 		first = 0;
-		sqlfile << "('" << region.code << "','" << region.active_only_mileage << "','" << region.active_preview_mileage << "')\n";
+		*fmt::format_to(fstr, "','{}','{}')\n", region.active_only_mileage, region.active_preview_mileage) = 0;
+		sqlfile << "('" << region.code << fstr;
 	}
 	sqlfile << ";\n";
 
@@ -293,7 +297,8 @@ void sqlfile1
 	    for (std::pair<Region* const,double>& rm : h.mileage_by_region)
 	    {	if (!first) sqlfile << ',';
 		first = 0;
-		sqlfile << "('" << h.systemname << "','" << rm.first->code << "','" << rm.second << "')\n";
+		*fmt::format_to(fstr, "','{}')\n", rm.second) = 0;
+		sqlfile << "('" << h.systemname << "','" << rm.first->code << fstr;
 	    }
 	sqlfile << ";\n";
 
@@ -313,7 +318,8 @@ void sqlfile1
 		first = 0;
 		auto it = t.active_only_mileage_by_region.find(rm.first);
 		double active_miles = (it != t.active_only_mileage_by_region.end()) ? it->second : 0;
-		sqlfile << "('" << rm.first->code << "','" << t.traveler_name << "','" << active_miles << "','" << rm.second << "')\n";
+		*fmt::format_to(fstr, "','{}','{}')\n", active_miles, rm.second) = 0;
+		sqlfile << "('" << rm.first->code << "','" << t.traveler_name << fstr;
 	  }
 	sqlfile << ";\n";
 
@@ -334,7 +340,8 @@ void sqlfile1
 		for (auto& rm : csmbr.second)
 		{	if (!first) sqlfile << ',';
 			first = 0;
-			sqlfile << "('" << systemname << "','" << rm.first->code << "','" << t.traveler_name << "','" << rm.second << "')\n";
+			*fmt::format_to(fstr, "{}", rm.second) = 0;
+			sqlfile << "('" << systemname << "','" << rm.first->code << "','" << t.traveler_name << "','" << fstr << "')\n";
 		}
 	  }
 	sqlfile << ";\n";
@@ -354,8 +361,9 @@ void sqlfile1
 		for (auto& rm : t.ccr_values)
 		{	if (!first) sqlfile << ',';
 			first = 0;
+			*fmt::format_to(fstr, "{}", rm.second) = 0;
 			sqlfile << "('" << rm.first->roots[0]->root << "','" << t.traveler_name << "','"
-				<< rm.second << "','" << (rm.second == rm.first->mileage) << "')\n";
+				<< fstr << "','" << (rm.second == rm.first->mileage) << "')\n";
 		}
 		sqlfile << ";\n";
 	}
@@ -374,8 +382,9 @@ void sqlfile1
 		for (std::pair<Route*,double>& rm : t.cr_values)
 		{	if (!first) sqlfile << ',';
 			first = 0;
+			*fmt::format_to(fstr, "{}", rm.second) = 0;
 			sqlfile << "('" << rm.first->root << "','" << t.traveler_name << "','"
-				<< rm.second << "','" << (rm.second >= rm.first->mileage) << "')\n";
+				<< fstr << "','" << (rm.second >= rm.first->mileage) << "')\n";
 		}
 		sqlfile << ";\n";
 	}

--- a/siteupdate/cplusplus/functions/tmstring.cpp
+++ b/siteupdate/cplusplus/functions/tmstring.cpp
@@ -1,3 +1,5 @@
+#define FMT_HEADER_ONLY
+#include <fmt/format.h>
 #include "tmstring.h"
 
 bool sort_1st_csv_field(const std::string& a, const std::string& b)
@@ -63,6 +65,15 @@ const char* strdstr(const char* h, const char* n, const char d)
 		  else if (!*b) return h;
 	}
 	return 0;
+}
+
+char* format_clinched_mi(char* str, double clinched, double total)
+{	/* return a nicely-formatted string for a given number of miles
+	clinched and total miles, including percentage */
+	if (total)
+		*fmt::format_to(str, "{:.2f} of {:.2f} mi ({:.2f}%)", clinched, total, 100*clinched/total) = 0;
+	else	*fmt::format_to(str, "{:.2f} of {:.2f} mi -.--%", clinched, total) = 0;
+	return str;
 }
 
 std::string double_quotes(std::string str)

--- a/siteupdate/cplusplus/functions/tmstring.h
+++ b/siteupdate/cplusplus/functions/tmstring.h
@@ -7,4 +7,5 @@ const char* upper(const char*);
 bool valid_num_str(const char*, const char);
 int strdcmp(const char*, const char*, const char);
 const char* strdstr(const char*, const char*, const char);
+char* format_clinched_mi(char*, double, double);
 std::string double_quotes(std::string);

--- a/siteupdate/cplusplus/tasks/threaded/StatsCsv.cpp
+++ b/siteupdate/cplusplus/tasks/threaded/StatsCsv.cpp
@@ -1,28 +1,20 @@
       #ifdef threading_enabled
 	HighwaySystem::it = HighwaySystem::syslist.begin();
-	switch(Args::mtcsvfiles ? Args::numthreads : 1)
-	{   case 1:
+	if (Args::numthreads == 1 || Args::stcsvfiles)
       #endif
-		cout << et.et() << "Writing allbyregionactiveonly.csv." << endl;
+	     {	cout << et.et() << "Writing allbyregionactiveonly.csv." << endl;
 		allbyregionactiveonly(0, active_only_miles);
 		cout << et.et() << "Writing allbyregionactivepreview.csv." << endl;
 		allbyregionactivepreview(0, active_preview_miles);
 		cout << et.et() << "Writing per-system stats csv files." << endl;
 		for (HighwaySystem& h : HighwaySystem::syslist) h.stats_csv();
+	     }
       #ifdef threading_enabled
-		break;
-	   case 2:
-		thr[0] = thread(allbyregionactiveonly,    &list_mtx, active_only_miles);
+	else {	thr[0] = thread(allbyregionactiveonly,    &list_mtx, active_only_miles);
 		thr[1] = thread(allbyregionactivepreview, &list_mtx, active_preview_miles);
-		thr[0].join();
-		thr[1].join();
-		break;
-	   default:
-		thr[0] = thread(allbyregionactiveonly,    nullptr, active_only_miles);
-		thr[1] = thread(allbyregionactivepreview, nullptr, active_preview_miles);
-		thr[2] = thread(StatsCsvThread, 2, &list_mtx);
-		thr[0].join();
-		thr[1].join();
-		thr[2].join();
-	}
+		// start at t=2, because allbyregionactive* will spawn another StatsCsvThread when finished
+		for (unsigned int t = 2; t < thr.size(); t++) thr[t] = thread(StatsCsvThread, t, &list_mtx);
+		THREADLOOP thr[t].join();
+	     }
       #endif
+

